File: src/spark/jogodavida_spark.scala
import org.apache.spark.sql.SparkSession

object JogoDaVidaSpark {

  // Função para calcular a próxima geração de uma única célula
  def computeNextState(currentState: Int, neighbors: Int): Int = {
    if (currentState == 1 && (neighbors < 2 || neighbors > 3)) 0
    else if (currentState == 0 && neighbors == 3) 1
    else currentState
  }

  // Função para calcular o número de vizinhos vivos para uma célula (i, j)
  def countNeighbors(grid: Array[Array[Int]], i: Int, j: Int, tam: Int): Int = {
    var neighbors = 0
    for (y <- i - 1 to i + 1) {
      for (x <- j - 1 to j + 1) {
        if ((y != i || x != j) && y >= 0 && y < tam && x >= 0 && x < tam) {
          neighbors += grid(y)(x)
        }
      }
    }
    neighbors
  }

  def main(args: Array[String]): Unit = {
    if (args.length < 2) {
      println("Uso: Jogo da Vida Spark <POWMIN> <POWMAX>")
      sys.exit(1)
    }

    val powMin = args(0).toInt
    val powMax = args(1).toInt
    
    val spark = SparkSession.builder.appName("Jogo da Vida Spark").getOrCreate()
    val sc = spark.sparkContext
    
    println(s"Executando Spark Engine para POWMIN=$powMin até POWMAX=$powMax")
    println("========================================")

    for (pow <- powMin to powMax) {
      val tam = 1 << pow
      val generations = 2 * (tam - 3)
      
      // Inicializa o tabuleiro (Grid)
      var grid = Array.fill(tam, tam)(0)
      grid(0)(1) = 1 // Padrão para validação
      grid(1)(2) = 1
      grid(2)(0) = 1
      grid(2)(1) = 1
      grid(2)(2) = 1

      val startTime = System.nanoTime()

      // Loop principal de gerações
      for (gen <- 1 to generations) {
        val gridRdd = sc.parallelize(0 until tam)

        // Mapeia cada linha para seu novo estado
        val nextGridRdd = gridRdd.map { i =>
          val newRow = Array.fill(tam)(0)
          for (j <- 0 until tam) {
            val neighbors = countNeighbors(grid, i, j, tam)
            newRow(j) = computeNextState(grid(i)(j), neighbors)
          }
          (i, newRow)
        }
        
        // Coleta os resultados e atualiza o grid principal
        val nextGridMap = nextGridRdd.collectAsMap()
        for ((i, row) <- nextGridMap) {
          grid(i) = row
        }
      }

      val endTime = System.nanoTime()
      val durationSeconds = (endTime - startTime) / 1e9d
      
      // Validação e Impressão de Resultados (similar ao código C)
      val finalLiveCells = grid.map(_.sum).sum
      println(s"tam=$tam; cells=$finalLiveCells; comp=${"%.7f".format(durationSeconds)} s")
    }

    spark.stop()
  }
}

File: src/core/jogodavida_spark.py
import sys
import time
from pyspark.sql import SparkSession

def wall_time():
    return time.time()

def ind2d(i, j, tam):
    return i * (tam + 2) + j

def UmaVida(tabulIn, tabulOut, tam):
    for i in range(1, tam + 1):
        for j in range(1, tam + 1):
            vizviv = (tabulIn[ind2d(i-1, j-1, tam)] + tabulIn[ind2d(i-1, j, tam)] +
                     tabulIn[ind2d(i-1, j+1, tam)] + tabulIn[ind2d(i, j-1, tam)] +
                     tabulIn[ind2d(i, j+1, tam)] + tabulIn[ind2d(i+1, j-1, tam)] +
                     tabulIn[ind2d(i+1, j, tam)] + tabulIn[ind2d(i+1, j+1, tam)])

            if tabulIn[ind2d(i, j, tam)] and vizviv < 2:
                tabulOut[ind2d(i, j, tam)] = 0
            elif tabulIn[ind2d(i, j, tam)] and vizviv > 3:
                tabulOut[ind2d(i, j, tam)] = 0
            elif not tabulIn[ind2d(i, j, tam)] and vizviv == 3:
                tabulOut[ind2d(i, j, tam)] = 1
            else:
                tabulOut[ind2d(i, j, tam)] = tabulIn[ind2d(i, j, tam)]

def InitTabul(tabulIn, tabulOut, tam):
    for ij in range((tam + 2) * (tam + 2)):
        tabulIn[ij] = 0
        tabulOut[ij] = 0
    
    tabulIn[ind2d(1, 2, tam)] = 1
    tabulIn[ind2d(2, 3, tam)] = 1
    tabulIn[ind2d(3, 1, tam)] = 1
    tabulIn[ind2d(3, 2, tam)] = 1
    tabulIn[ind2d(3, 3, tam)] = 1

def Correto(tabul, tam):
    cnt = 0
    
    for ij in range((tam + 2) * (tam + 2)):
        cnt = cnt + tabul[ij]
    
    return (cnt == 5 and tabul[ind2d(tam-2, tam-1, tam)] and
            tabul[ind2d(tam-1, tam, tam)] and tabul[ind2d(tam, tam-2, tam)] and
            tabul[ind2d(tam, tam-1, tam)] and tabul[ind2d(tam, tam, tam)])

def main(powmin, powmax):
    spark = SparkSession.builder \
        .appName("JogoDaVidaSparkDireto") \
        .master("local[*]") \
        .getOrCreate()
    spark.sparkContext.setLogLevel("ERROR")
    
    print("Executando Jogo da Vida Spark")
    print("==========================================================")
    
    try:
        for pow in range(powmin, powmax + 1):
            tam = 1 << pow

            t0 = wall_time()
            tabulIn = [0] * ((tam + 2) * (tam + 2))
            tabulOut = [0] * ((tam + 2) * (tam + 2))
            InitTabul(tabulIn, tabulOut, tam)
            t1 = wall_time()
            
            for i in range(2 * (tam - 3)):
                UmaVida(tabulIn, tabulOut, tam)
                UmaVida(tabulOut, tabulIn, tam)
            
            t2 = wall_time()
            
            if Correto(tabulIn, tam):
                print("**Ok, RESULTADO CORRETO**")
            else:
                print("**Nok, RESULTADO ERRADO**")
            
            t3 = wall_time()
            
            print(f"tam={tam}; tempos: init={t1-t0:.7f}, comp={t2-t1:.7f}, fim={t3-t2:.7f}, tot={t3-t0:.7f}")
    
    finally:
        spark.stop()

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Uso: jogodavida_spark.py <powmin> <powmax>", file=sys.stderr)
        sys.exit(-1)
    main(int(sys.argv[1]), int(sys.argv[2]))